src/librbd/mirror/snapshot/CreatePrimaryRequest.cc

```diff
@@ -194,6 +194,10 @@ void CreatePrimaryRequest<I>::unlink_peer() {
     size_t count = 0;
     uint64_t unlink_snap_id = 0;
     for (auto &snap_it : m_image_ctx->snap_info) {
+      // skip UnlinkPeerRequest for recent mirror snapshot
+      if (snap_it.first == *m_snap_id) {
```

Comment:
I spent the evening looking into this and I came up with fix: https://github.com/idryomov/ceph/commit/13b78573b4e3fc56cdef22033168aacfc59b1123
Please try it out, including injecting a crash as before as that is the only actual reproducer we have.

Reply:
> m_snap_id may be nullptr

@idryomov 

Ouch! I completely forget about this while pushing the changes, my actual testing changes take care about nullptr (tested on a different machine), something like:

```
// skip UnlinkPeerRequest for recent mirror snapshot
if (m_snap_id != nullptr && snap_it.first == *m_snap_id) {
  continue;
}
```

I think the below is there for a reason, If you don't see any issue with above, can we live with it without discarding below?
```
 // if snapshot created via remote RPC, refresh is required to retrieve
  // the snapshot id
  if (m_snap_id == nullptr) {
    unlink_peer();
    return;
  }
```

Reply:
That might be fixed like:
```
pkalever@localhost ceph]$ git diff src/librbd/mirror/snapshot/CreatePrimaryRequest.cc
diff --git a/src/librbd/mirror/snapshot/CreatePrimaryRequest.cc b/src/librbd/mirror/snapshot/CreatePrimaryRequest.cc
index 36cd3bddfc7..a47de024f23 100644
--- a/src/librbd/mirror/snapshot/CreatePrimaryRequest.cc
+++ b/src/librbd/mirror/snapshot/CreatePrimaryRequest.cc
@@ -193,7 +193,13 @@ void CreatePrimaryRequest<I>::unlink_peer() {
     std::shared_lock image_locker{m_image_ctx->image_lock};
     size_t count = 0;
     uint64_t unlink_snap_id = 0;
+    auto max_snaps = max_snapshots;
     for (auto &snap_it : m_image_ctx->snap_info) {
+      // skip UnlinkPeerRequest for recent mirror snapshot
+      if (m_snap_id != nullptr && snap_it.first == *m_snap_id) {
+        max_snaps--;
+        continue;
+      }
       auto info = std::get_if<cls::rbd::MirrorSnapshotNamespace>(
         &snap_it.second.snap_namespace);
       if (info == nullptr) {
@@ -209,19 +215,20 @@ void CreatePrimaryRequest<I>::unlink_peer() {
       // or if it's not linked with any peer (happens if mirroring is enabled
       // on a pool with no peers configured or if UnlinkPeerRequest gets
       // interrupted)
-      if (info->mirror_peer_uuids.size() == 0) {
+      if (!info->mirror_peer_uuids.empty() &&
+          info->mirror_peer_uuids.count(peer) == 0) {
+        continue;
+      }
+      if (info->mirror_peer_uuids.empty() || !info->complete) {
         peer_uuid = peer;
         snap_id = snap_it.first;
         break;
       }
-      if (info->mirror_peer_uuids.count(peer) == 0) {
-        continue;
-      }
       count++;
-      if (count == max_snapshots) {
+      if (count == max_snaps) {
         unlink_snap_id = snap_it.first;
       }
-      if (count > max_snapshots) {
+      if (count > max_snaps) {
         peer_uuid = peer;
         snap_id = unlink_snap_id;
         break;
[pkalever@localhost ceph]$ 
```


But with your proposed fix aren't we missing to call `unlink_peer()` itself in `void CreatePrimaryRequest<I>::refresh_image()` when `m_snap_id == nullptr` ?

Reply:
> That might be fixed like:

Did you test that at all?  Besides leaving `m_snap_id == nullptr` case unhandled, it quite obviously breaks `rbd_mirroring_max_mirroring_snapshots` logic because `snap_it.first == *m_snap_id` is likely to be true for the last iteration (i.e. newest snapshot), meaning that `max_snaps` would be decremented right before bailing out of the entire loop which makes the whole change useless.


> But with your proposed fix aren't we missing to call `unlink_peer()` itself in `void CreatePrimaryRequest<I>::refresh_image()` when `m_snap_id == nullptr` ?

We aren't -- `unlink_peer()` is called unconditionally before returning from `refresh_image()` there.

Reply:
> We aren't -- unlink_peer() is called unconditionally before returning from refresh_image() there.
Ah! didn't notice this one (sorry for the mess, was switching between multiple things)
Thanks will align to the fix your proposed and test it a bit and send it.